Firestore/Example/Tests/Integration/API/FIRDatabaseTests.m

```diff
@@ -931,4 +932,175 @@ - (void)testWatchStreamReconnectsAfterIdle {
+  XCTAssertFalse(result.metadata.fromCache);
+  XCTAssertEqualObjects(result.data, initialData);
+}
+
+- (void)testGetDocumentWhileOfflineWithDefaultGetOptions {
+  FIRDocumentReference *doc = [self documentRef];
+
+  // set document to a known value
+  NSDictionary<NSString *, id> *initialData = @{@"key1" : @"value1"};
+  [self writeDocumentRef:doc data:initialData];
+
+  // go offline for the rest of this test
+  [self disableNetwork];
+
+  // update the doc (though don't wait for a server response. We're offline; so
+  // that ain't happening!). This allows us to further distinguished cached vs
+  // server responses below.
+  NSDictionary<NSString *, id> *newData = @{@"key2" : @"value2"};
+  [doc setData:newData completion:^(NSError *_Nullable error) {
+    XCTAssertTrue(false, "Because we're offline, this should never occur.");
+  }];
+
+  // get doc and ensure it exists, *is* from the cache, and matches the
+  // newData.
+  FIRDocumentSnapshot *result = [self readDocumentForRef:doc];
+  XCTAssertTrue(result.exists);
+  XCTAssertTrue(result.metadata.fromCache);
+  XCTAssertEqualObjects(result.data, newData);
+}
+
+- (void)testGetDocumentWhileOnlineCacheOnly {
+  FIRDocumentReference *doc = [self documentRef];
+
+  // set document to a known value
+  NSDictionary<NSString *, id> *initialData = @{@"key" : @"value"};
+  [self writeDocumentRef:doc data:initialData];
+
+  // get doc and ensure that it exists, *is* from the cache, and matches
+  // the initialData.
+  FIRDocumentSnapshot *result = [self readDocumentForRef:doc getOptions:[FIRGetOptions fromCache]];
+  XCTAssertTrue(result.exists);
+  XCTAssertTrue(result.metadata.fromCache);
+  XCTAssertEqualObjects(result.data, initialData);
+}
+
+- (void)testGetDocumentWhileOfflineCacheOnly {
+  FIRDocumentReference *doc = [self documentRef];
+
+  // set document to a known value
+  NSDictionary<NSString *, id> *initialData = @{@"key1" : @"value1"};
+  [self writeDocumentRef:doc data:initialData];
+
+  // go offline for the rest of this test
+  [self disableNetwork];
+
+  // update the doc (though don't wait for a server response. We're offline; so
+  // that ain't happening!). This allows us to further distinguished cached vs
+  // server responses below.
+  NSDictionary<NSString *, id> *newData = @{@"key2" : @"value2"};
+  [doc setData:newData completion:^(NSError *_Nullable error) {
+    XCTAssertTrue(false, "Because we're offline, this should never occur.");
+  }];
+
+  // get doc and ensure it exists, *is* from the cache, and matches the
+  // newData.
+  FIRDocumentSnapshot *result = [self readDocumentForRef:doc getOptions:[FIRGetOptions fromCache]];
+  XCTAssertTrue(result.exists);
+  XCTAssertTrue(result.metadata.fromCache);
+  XCTAssertEqualObjects(result.data, newData);
+}
+
+- (void)testGetDocumentWhileOnlineServerOnly {
+  FIRDocumentReference *doc = [self documentRef];
+
+  // set document to a known value
+  NSDictionary<NSString *, id> *initialData = @{@"key" : @"value"};
```

Comment:
Can you make initialData and updateData a member of this test class so you can access it from everywhere?

All your tests could also be shorter if you only asserted on the properties you are testing.